File: javascript/lca.js
function findPath( root, val1, path ) {
   if ( path == null ) path = [];
   if ( root == null ) return null;

   path.push(root);
   if ( root.val == val1 ) return path;
   if ( root.val < val1 ) return findPath( root.left, val1, path );
   if ( root.val > val1 ) return findPath( root.right, val1, path );
}

function findLca( root, val1, val2){
   var p1 = findPath(root,val1);
   var p2 = findPath(root,val2); 

   if ( p1 == null || p2 == null ) return null;
   for ( var i = 0, j = 0; i < p1.length, j < p2.length; ++i, ++j ) {
       if ( p1[i].val != p2[j].val ) break;
   }

   if ( i == 0 ) return null;
   return p1[i-1];
}

File: java/src/LCA.java
import java.util.Deque;
import java.util.LinkedList;

/*
 LCA

 return -1 if not found

O(N)

 */
public class LCA {

    // Graph node class
    public static class Node {
        private final int value;
        Node left;
        Node right;
        public Node(int value) {
            this.value = value;
        }
    }

    private static boolean dfs( Node node, Integer val, Deque<Node> dq ){
        if ( node == null ) return false;
        dq.add(node);
        if ( node.value == val ) return true;
        if ( dfs(node.left,val,dq) ) return true;
        if ( dfs(node.right,val,dq) ) return true;
        dq.removeLast();
        return false;
    }

    public static Integer find( Node root, Integer val1, Integer val2){
        Deque<Node> dq1 = new LinkedList<>();
        if ( !dfs(root,val1,dq1) ) return -1;
        Deque<Node> dq2 = new LinkedList<>();
        if ( !dfs(root,val2,dq2) ) return -1;

        int lca = root.value;
        while( !dq1.isEmpty() && !dq2.isEmpty() && dq1.peek().equals(dq2.peek()) ){
            lca = dq1.peek().value;
            dq1.remove();
            dq2.remove();
        }
        return lca;
    }

     public static void main(String[] args){

         /* Construct tree
                  1
               /    \
             2       3
            / \     / \
           4   5   6   7
         */
         Node root = new Node(1);
         root.left = new Node(2);
         root.right = new Node(3);
         root.left.left = new Node(4);
         root.left.right = new Node(5);
         root.right.left = new Node(6);
         root.right.right = new Node(7);

         System.out.println( "lca(4,5): " + find(root,4,5));
         System.out.println( "lca(1,5): " + find(root,1,5));
     }

} 